Volume/Supporting Views/RefreshableScrollView.swift

```diff
@@ -0,0 +1,135 @@
+
+    static var defaultValue = [Position]()
+
+    static func reduce(value: inout [Position], nextValue: () -> [Position]) {
+        value.append(contentsOf: nextValue())
+    }
+}
+
+private struct PositionIndicator: View {
+    let type: PositionType
+
+    var body: some View {
+        GeometryReader { proxy in
+            Color.clear
+                .preference(
+                    key: PositionPreferenceKey.self,
+                    value: [Position(type: type, y: proxy.frame(in: .global).minY)]
+                )
+        }
+    }
+}
+
+struct RefreshableScrollView<Content: View>: View {
+    typealias RefreshComplete = () -> Void
+    typealias OnRefresh = (@escaping RefreshComplete) -> Void
+
+    @State private var state = RefreshState.waiting // the current state
+    private let THRESHOLD: CGFloat = 50
+    
+    let content: Content
+    let onRefresh: OnRefresh
+
+    private enum RefreshState {
+      case waiting, primed, loading
+    }
+    
+    init(onRefresh: @escaping OnRefresh, @ViewBuilder content: () -> Content) {
+        self.onRefresh = onRefresh
+        self.content = content()
+    }
+
+    var body: some View {
+        ScrollView(showsIndicators: false) {
+            ZStack(alignment: .top) {
+                PositionIndicator(type: .moving)
+                    .frame(height: 0)
+
+                content
+                    .alignmentGuide(.top) { _ in
+                        (state == .loading) ? -THRESHOLD : 0
+                    }
+
+                ZStack {
+                    Rectangle()
+                        .foregroundColor(.clear)
+                        .frame(height: THRESHOLD)
+                    ActivityIndicator(isAnimating: state == .loading) {
+                        $0.hidesWhenStopped = false
+                    }
+                }
+                .offset(y: (state == .loading) ? 0 : -THRESHOLD)
+            }
+        }
+        .background(PositionIndicator(type: .fixed))
+        .onPreferenceChange(PositionPreferenceKey.self) { values in
+            if state != .loading {
+                DispatchQueue.main.async {
+                    let movingY = values.first { $0.type == .moving }?.y ?? 0
+                    let fixedY = values.first { $0.type == .fixed }?.y ?? 0
+                    let offset = movingY - fixedY
+
+                    if offset > THRESHOLD && state == .waiting {
+                        state = .primed
+                    } else if offset < THRESHOLD && state == .primed {
+                        state = .loading
+                        onRefresh {
+                            withAnimation {
+                                self.state = .waiting
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+struct ActivityIndicator: UIViewRepresentable {
+    public typealias UIView = UIActivityIndicatorView
+    public var isAnimating: Bool = true
```

Comment:
Delete `public`. Also, SwiftUI has a built in activity indicator. It's called a `ProgressView` with a `CircularProgressViewStyle`. Example [here](https://stackoverflow.com/questions/56496638/activity-indicator-in-swiftui). You probably don't need any of the code below line 114.